pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -520,18 +572,53 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
 //  fqdn/NameManager instance).
 //  - Write the response to the endpoint.
 func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
-	stat := ProxyRequestContext{}
-	stat.ProcessingTime.Start()
 	requestID := request.Id // keep the original request ID
 	qname := string(request.Question[0].Name)
 	protocol := w.LocalAddr().Network()
+	epIPPort := w.RemoteAddr().String()
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.DNSName:      qname,
-		logfields.IPAddr:       w.RemoteAddr(),
-		logfields.DNSRequestID: request.Id})
+		logfields.IPAddr:       epIPPort,
+		logfields.DNSRequestID: requestID,
+	})
+
+	var stat ProxyRequestContext
+	if p.ConcurrencyLimit != nil {
+		// TODO: Consider plumbing the daemon context here.
+		ctx, cancel := context.WithTimeout(context.TODO(), p.ConcurrencyGracePeriod)
+		defer cancel()
+
+		stat.SemaphoreAcquireTime.Start()
+		if p.ConcurrencyGracePeriod == 0 {
+			// No grace time configured. Failing to acquire semaphore means
+			// immediately give up.
+			if !p.ConcurrencyLimit.TryAcquire(1) {
+				errSema := ErrFailedAcquireSemaphore{
+					parallel: option.Config.DNSProxyConcurrencyLimit,
+				}
+				p.handleFailedSemaphoreAcquire(scopedLog, &stat, errSema, epIPPort, protocol, w, request)
```

Comment:
This implementation looks correct to me.

I have a minor concern for readability / confirming that this implementation is correct: For most error conditions in this function `ServeDNS()`, they directly invoke all of the error handling for stats time updates &errors, send the monitor messages, and sending DNS error responses. So for instance just by reading this function, we can see that `stat.ProcessingTime` gets started, and then if you scan down to all of the `return` statements, you can confirm that this function also ends the `stat.ProcessingTime`. When the `Start` and `End` are not in the same function for the `SemaphoreAcquireTime` as is currently proposed, it becomes much easier to accidentally break them or extend the code in a way that ignores this call "balancing", because the details are hidden underneath `p.handleFailedSemaphoreAcquire()`. Likewise it's harder to verify that the other calls like `p.sendRefused()` are consistently handled in all error cases.

While I think it's definitely tidier to combine several repeated lines together in a function, I actually wonder if it would be more readable to factor this entire semaphore handling piece into a child function which returns success/failure depending on the overall semaphore handling logic. The whole `if p.ConcurrencyLimit != nil` condition could instead be something like:

```
stat.SemaphoreAcquireTime.Start()
if err := p.EnforceConcurrencyLimit(); err != nil {
	stat.SemaphoreAcquireTime.End(false)
	if p.logLimiter.Allow() {
		scopedLog.WithError(errToPass).Error("Dropping DNS request due to too many DNS requests already in-flight")
	}
	stat.Err = err
	p.NotifyOnDNSMsg(time.Now(), nil, epIPPort, 0, "", request, protocol, false, stat)
	p.sendRefused(scopedLog, w, request)
        return
}
stat.SemaphoreAcquireTime.End(true)
```

Then, the matching start/end and DNS response error handling logic would be consistent with the other error cases in this function (which is a simple way to check if it's correct), and the size of this function will not grow so much. The exact details of TryAcquire() / Acquire(), context and error preparation could be hidden inside the `EnforceConcurrencyLimit()` function. There'd be a separation between "Do we need to ratelimit?" (child function) and "what does `ServeDNS()` need to ensure happens during error conditions?" (always handled the same way for all error conditions in `ServeDNS()`.)

Just a thought, doesn't have to be this way but I think it could make this code less error-prone to further extend in future.

Reply:
Ah, this is a great suggestion. Let me give it a shot.

Reply:
^ Could potentially even pass `stat` into the `EnforceConcurrencyLimit()` so that this function doesn't have to handle the `stat.SemaphoreAcquireTime`, but then the rest of the function is handling all of the `stat` accounting so it probably makes more sense for `stat` interactions to stay in this function. This way it's clearer that `ServeDNS()` handles all of the `stat` updates and then the child functions don't need to worry about this.

Reply:
Implemented.

Reply:
Yeah seems a bit cleaner to keep `stat` handling inside the parent function. Less side effects.